database/reportdb.go

```diff
@@ -0,0 +1,147 @@
+package database
+
+import (
+	"context"
+	"time"
+
+	lock "github.com/square/mongo-lock"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+const (
+	// resourceReports is the resource name used when locking reports
+	resourceReports = "ncmec_reports"
+)
+
+type (
+	// NCMECReport is a database entity that represents an NCMEC report.
+	NCMECReport struct {
+		ID      primitive.ObjectID `bson:"_id"`
+		EmailID primitive.ObjectID `bson:"email_id"`
+
+		Filed    bool      `bson:"filed"`
+		FiledAt  time.Time `bson:"filed_at"`
+		FiledErr string    `bson:"filed_err"`
+
+		Report   string `bson:"report"`
+		ReportID uint64 `bson:"report_id"`
+
+		InsertedAt time.Time `bson:"inserted_at"`
+	}
+
+	// ncmecReportLock represents a lock on an ncmec report.
+	ncmecReportLock struct {
+		staticClient         *lock.Client
+		staticEmailUID       string
+		staticPortalHostname string
+	}
+)
+
+// NewReportLock returns a lock on a report entity
+func (db *AbuseScannerDB) NewReportLock(reportID string) *abuseLock {
+	return db.newLockCustom(resourceReports, reportID)
+}
+
+// InsertReport will try and insert the given report into the database.
+func (db *AbuseScannerDB) InsertReport(report NCMECReport) error {
+	ctx, cancel := context.WithTimeout(context.Background(), mongoDefaultTimeout)
+	defer cancel()
+
+	coll := db.staticDatabase.Collection(collNCMECReports)
+	_, err := coll.InsertOne(ctx, report)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// FindReport returns the report for given object id.
+func (db *AbuseScannerDB) FindReport(reportID primitive.ObjectID) (*NCMECReport, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), mongoDefaultTimeout)
+	defer cancel()
+
+	coll := db.staticDatabase.Collection(collNCMECReports)
+	res := coll.FindOne(ctx, bson.M{"_id": reportID})
+	if isDocumentNotFound(res.Err()) {
+		return nil, nil
```

Comment:
Yeah, this is the thing we go back and forth on right 😛  I always suggest you do it this way in your PRs. I like this personally, if you get a pointer you should generally always check if it's nil or not and I like the fact that you, as a caller, don't have to think about database-specific errors like the mongo err no documents.

I know we could return a custom `ErrNoDocument` but at this point it would be inconsistent to handle it this way seeing as in all of the abuse repos I use the same pattern. I don't mind updating it if I stand alone on this but I'll do it in a F/U to tackle all occurrences.

But give it a chance, I think it's pretty clean and it reads nice and it's one less error to handle and you get around the awkward pattern of checking an error, seeing it's not nil and going "oh no", and then "oof" it's just an `ErrNoDocuments`... I mean you're getting a pointer, to me this is the equivalent of a `Promise<object|null>` in javascript for instance.

Reply:
Oh, haha, it's true that we've been over this. It would be nice to get some input from other team members here. cc @ChrisSchinnerl @MSevey 

I'm not convinced this approach is less work because I check for errors anyway and now I'll need to check for `nil` as well. I don't like that but I'm not some kind of hill to die on, so if the team likes the `nil` approach I can start using it. I agree that passing DB errors leads to leaky abstractions and that custom `not found` errors are a bit clumsy because you need to define them (I have a bunch of them in `accounts`).

Reply:
I'd stick to the conventions. e.g. `os` exposes `os.IsNotExist`. In this case I'd expose a `database.IsNotFound` that you can use.

This is clearer than knowing you have to check for `nil` even if there is no error. It's also not more lines of code. When you know there must be a document it's actually less verbose.